src/protocols/bdx/BDXMessageUtils.cpp

```diff
@@ -0,0 +1,432 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Implements utility methods for working with some complex BDX messages.
+ */
+
+#include <protocols/bdx/BDXMessageUtils.h>
+
+#include <support/BufBound.h>
+#include <support/BufferReader.h>
+#include <support/CodeUtils.h>
+
+#include <limits>
+
+#define VERSION_MASK 0x0F
+#define SENDER_DRIVE_MASK 0x10
+#define RECEIVER_DRIVE_MASK 0x20
+#define ASYNC_MASK 0x40
+
+#define DEFLEN_MASK 0x01
+#define START_OFFSET_MASK 0x02
+#define WIDERANGE_MASK 0x10
+
+using namespace chip;
+using namespace chip::BDX;
+using namespace chip::Encoding::LittleEndian;
+
+CHIP_ERROR TransferInit::Pack(System::PacketBuffer & aBuffer) const
+{
+    CHIP_ERROR err              = CHIP_NO_ERROR;
+    uint8_t rangeCtl            = 0;
+    uint8_t proposedTransferCtl = 0;
+    bool widerange              = ((mStartOffset | mMaxLength) > std::numeric_limits<uint32_t>::max());
```

Comment:
It's a shortcut, but is it wrong? One way to look at it is that we only care if either of the values have a bit set outside of the uint32_t range. If that's true, the result will be greater than UINT32_MAX. If neither are greater than UINT32_MAX, then they won't have a bit set in the high range, and this check will return false.

Reply:
Not wrong in this particular case (since we are talking about 0xFFF.... values), but it does result in a double-take when reading code of 'is it really correct?'.

I would err towards easier to read/understand code since is is unlikely that we need some potential performance boost here.

Reply:
Ok, I was trying to reduce the length of the expression but I understand how this could be confusing. I'll take Tennessee's suggestion